### PR TITLE
Fix Bootstrap password prompt

### PR DESCRIPTION
**Pull request by samshinde**
Bootstrap command was not prompting for the password when --connection-password option is not passed.

## Description
 
Bootstrap command was throwing error `ERROR: RuntimeError: password is a required option` when --connection-password option is not passed. The exception was not getting raised because of an incorrect method call.
- Done correct exception handling
- Fixed test case

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/knife/bootstrap.rb [hunk 1]
```diff
@@ -614,9 +614,9 @@ def perform_bootstrap(remote_bootstrap_script_path)
         end
       end
 
-      def connect!
+      def connect!(opts = nil)
```
Comment (Nimesh-Msys) [inline lib/chef/knife/bootstrap.rb]: Any thoughts to [Keyword Arguments](https://thoughtbot.com/blog/ruby-2-keyword-arguments) ?

File: lib/chef/knife/bootstrap.rb [hunk 2]
```diff
@@ -638,8 +638,10 @@ def connect!
           EOM
           # FIXME: this should save the key to known_hosts but doesn't appear to be
           config[:ssh_verify_host_key] = :accept_new
-          do_connect(connection_opts(reset: true))
-        elsif ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed
+          conn_opts = connection_opts(reset: true)
+          opts.merge! conn_opts
+          connect!(opts)
+        elsif ssh? && e.class == Train::ClientError && e.message == "Your SSH Agent has no keys added, and you have not specified a password or a key file"
```
Comment (Nimesh-Msys) [inline lib/chef/knife/bootstrap.rb]: Error check with string comparison could be risky. Also @ L#660, would you please confirm if this is really the last option.

File: lib/chef/knife/bootstrap.rb [hunk 3]
```diff
@@ -650,7 +652,23 @@ def connect!
           end
 
           opts.merge! force_ssh_password_opts(password)
-          do_connect(opts)
+          connect!(opts)
+        else
+          raise
```
Comment (Nimesh-Msys) [inline lib/chef/knife/bootstrap.rb]: Best if we could have redefine [exceptions](https://github.com/chef/chef/blob/master/lib/chef/exceptions.rb) and can raise and catch them.

File: lib/chef/knife/bootstrap.rb (hunk 2 above)
Comment (vsingh-msys) [reply lib/chef/knife/bootstrap.rb]: @Nimesh-Msys sure this might be an improvement area for https://github.com/inspec/train

File: lib/chef/knife/bootstrap.rb (hunk 3 above)
Comment (lamont-granquist) [reply lib/chef/knife/bootstrap.rb]: that bit is actually fine, the code is just stating that it handles certain Train:Error's but not others.

File: lib/chef/knife/bootstrap.rb [hunk 4]
```diff
@@ -638,8 +638,10 @@ def connect!
           EOM
           # FIXME: this should save the key to known_hosts but doesn't appear to be
           config[:ssh_verify_host_key] = :accept_new
-          do_connect(connection_opts(reset: true))
-        elsif ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed
+          conn_opts = connection_opts(reset: true)
+          opts.merge! conn_opts
```
Comment (marcparadise) [inline lib/chef/knife/bootstrap.rb]: Why the change to merge into old opts here?

File: lib/chef/knife/bootstrap.rb [hunk 5]
```diff
@@ -638,8 +638,10 @@ def connect!
...
+          retry
+        elsif (ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed) || (ssh? && e.class == Train::ClientError && e.message == "Your SSH Agent has no keys added, and you have not specified a password or a key file")
```
Comment (marcparadise) [inline lib/chef/knife/bootstrap.rb]: This would work, but I'd prefer updating train [1] to populate the exception's `reason` field so that we don't have to worry about changes to text. This'll also keep moving train towards providing a consistent way to discover exception reasons.
 
[1] https://github.com/inspec/train/blob/master/lib/train/transports/ssh.rb#L106
to something like: 
```
        raise Train::ClientError.new(
            "Your SSH Agent has no keys added, and you have not specified a password or a key file",
            :no_ssh_password_or_key_available
        )
```

File: lib/chef/knife/bootstrap.rb (hunk 4 above)
Comment (samshinde) [reply lib/chef/knife/bootstrap.rb]: Because if --connection-password is not passed and `Net::SSH::AuthenticationFailed` or `Train::ClientError` error is thrown then old opts will have the password.

File: lib/chef/knife/bootstrap.rb (hunk 5 above)
Comment (samshinde) [reply lib/chef/knife/bootstrap.rb]: Updated comparison with `e.reason`. Added PR https://github.com/inspec/train/pull/517 for this change.

**Comment by vsingh-msys** [comment]
@tas50 this PR is good to merge as inspec/train#517 is now in master. Also, need to pin train 3.1+
for the latest changes.
```
s.add_dependency "train-core", "~> 3.1" 
```
